### PR TITLE
Breakpoints on pipeline stages

### PR DESCRIPTION
**Pull request by dsyme**
This adds breakpoints and stepping on stages in F# forward piping in debug code.

Spec (debug codegen only)

### Breakpoint validity

For `expr |> f1 |> f2 ... |> fLast` , a breakpoint is valid on the expression that is an input to a pipeline, and any "stage" of a pipeline.

Similarly, for `(expr1, expr2) ||> f1 |> f2 ... |> fLast` and (expr1, expr2, expr3) |||> f1 |> f2 ... |> fLast` , a breakpoint is valid on each of the input expressions, and any "stage" of a pipeline.

### Sequence point emit 

A corresponding sequence point is emitted in .NET IL before the evaluation of `expr` and before applying each of `f1`, `f2` etc.

### Locals

*  A local value `Pipe #m input #n @ line` is shown within f1, f2, f3, ...for the corresponding value of `expr`. The pipe number is the occurrence of piping within the method.  The `n` it the input number (elided if only one input).  The `line` is the line number where the input expression occurs.

* A local value `Pipe #m stage 1 @ line` is shown within f2, f3, ...  for the corresponding value of `expr |> f1`

* A local value `Pipe #m stage 2 @ line` is shown within f3, ...  for the corresponding value of `expr |> f1 |> f2`

* etc.

The names of the locals are emitted in IL code and are not localized (as that would make compiler output dependent on localization).

Sample screen capture below (stepping) - the local values names differ slightly from the above spec

https://user-images.githubusercontent.com/7204669/129215988-7067c703-ca5f-44cc-af2f-33a11bef4890.mp4

And with breakpoints:

https://user-images.githubusercontent.com/7204669/129219214-4a146dd4-5d37-48c9-ba33-b06c5bf01a9a.mp4

### CODE REVIEW

File: src/fsharp/SyntaxTreeOps.fs [hunk 1]
```diff
@@ -759,4 +759,11 @@ let (|ParsedHashDirectiveArguments|) (input: ParsedHashDirectiveArgument list) =
...
         | ParsedHashDirectiveArgument.String (s, _, _) -> s
         | ParsedHashDirectiveArgument.SourceIdentifier (_, v, _) -> v)
-        input
\ No newline at end of file
+        input
+
+let (|SynPipeRight|_|) input =
+    match input with
+    | SynExpr.App (ExprAtomicFlag.NonAtomic, false, SynExpr.App (ExprAtomicFlag.NonAtomic, true, SynExpr.Ident synId, x1, _m1), x2, _m2) 
+        when synId.idText = "op_PipeRight" ->
```
Comment (cartermp) [inline src/fsharp/SyntaxTreeOps.fs:767]: Is there a reason to limit this to just single pipes?

Comment (Happypig375) [reply src/fsharp/SyntaxTreeOps.fs:767]: I'd like double-pipes and triple-pipes to be supported as well, also the back-pipes

Comment (dsyme) [reply src/fsharp/SyntaxTreeOps.fs:767]: > Is there a reason to limit this to just single pipes?

No, no reason, just needs a bit of work

> also the back-pipes

Back pipes will not be supported for this.  They should be used very rarely (see my talk "F# Code I love"), and are therefore not a priority for tooling like this.

Comment (cartermp) [reply src/fsharp/SyntaxTreeOps.fs:767]: Yes, double and triple could be reasonable, but I don't think backpipes are that good. The direction would not be top-down, left-to-right, unlike nearly every other statement that gets debugged.

**Comment by dsyme** [comment]
Some unresolved questions:

- This increases the number of locals we emit, e.g. is this a problem in methods with massive pipelines

- The same local names are used regardless of pipeline.  When there are sub-pipelines-within-pipelines this may cause confusion.   However it's probably a pricce worth paying.
 
- Breakpoints and sequence points suddenly appearing on innocuous uses of pipelines outside the statement structure, e.g. do we put down debug sequence and breakpoints for pipelines in argument position?  Currently in this PR we do.  That's OK I guess but may be a bit surprising when stepping. e.g. consider

  ```fsharp
  someFunction (x |> someThing) (x |> someOtherThing) 
  ```

**Comment by cartermp** [comment]
@dsyme you can debug creating a completion list as a decent test: https://github.com/dotnet/fsharp/blob/8a62ed9c3d96ba5595096841b15e3070c385a9b6/src/fsharp/service/ServiceDeclarationLists.fs#L956

**Comment by dsyme** [comment]
> As far as the names generated for locals in pipelines, IIRC the C# expression evaluator has issues with shadowed names, 

It does, though as things stand the names aren't actually usable in expression evaluation (`<pipe-input>` etc.). Not sure what to do about this.  If we choose usable names they will potentially conflict with user identifiers.  If we don't they can't be used in expression evaluation watch window etc.

> is it possible to generate unique names (or close-to-unique?) for the intermediate values to sidestep this?

We could in theory generate unique names (`pipe1-input`, `pipe1-stage1` etc?), though I do wonder if it would be more confusing.

**Comment by baronfel** [comment]
> the names aren't actually usable in expression evaluation

Great point. In VSCode at least (unsure about VS) method returns do have a local that's rendered something like:

`MEMBER_NAME returned: RETURN_VALUE`

so there might be an option there for us:

`PIPELINE_STAGE_TEXT yielded: STAGE_INTERMEDIATE_VALUE`

**Comment by Happypig375** [comment]
If the expression evaluator can support F#, then we can use \`\`pipe@input\`\` because `warning FS1104: Identifiers containing '@' are reserved for use in F# code generation`

**Comment by dsyme** [comment]
Folks, could you choose between these as names for generated locals please?  I need some guidance here:

### Option 1 - succinct

```
pipe1-input 
pipe1-stage1
pipe2-stage2
```

### Option 2 - usable in C# expression evaluator but confusing because they look like F# values

```
pipe1_input 
pipe1_stage1
pipe2_stage2
```

### Option 3 - uppercase start, spaces with text
```
Pipe #1 input
Pipe #1 stage #1
Pipe #1 stage #2
```

### Option 4 - something else, please spec

**Comment by auduchinok** [comment]
> If we choose usable names they will potentially conflict with user identifiers.

Can we generate usable names in a similar way to how type parameter names are generated, i.e. keep track of used names? Or, say, similar to function parameter names.

All local names are known at this stage of compilation, aren't they?

**Comment by dsyme** [comment]
> Can we generate usable names in a similar way to how type parameter names are generated, i.e. keep track of used names? Or, say, similar to function parameter names.

I mean, we could.  But I think a bigger problem is cognitive - people look at "pipe1_input" and think it's an F# identifier and wonder what the heck it is.  So I suspect that, like return values from functions, it's better to produce text like "Pipe #1 input" or "pipe1-input"

**Comment by baronfel** [comment]
I agree, I think the names should unambiguously not be F# identifiers if possible. So Option 3 would be my ideal of the ones presented so far.

**Comment by auduchinok** [comment]
Can this technique also be used in other expressions with implicit "parameters"? If yes, it'd be great to take cases like `function` or `try ... with` expressions into account when choosing naming scheme.

**Comment by auduchinok** [comment]
> But I think a bigger problem is cognitive - people look at "pipe1_input" and think it's an F# identifier and wonder what the heck it is.

Has it been problem for `arg1` and similar identifiers before? If not, I'd think it'll be fine here too.

On the other hand, I'd prefer being able to use these identifiers in evaluating expressions in debugger over anything else, so I'm slightly biased. I really think evaluating expressions is an important part here.

**Comment by dsyme** [comment]
> Has it been problem for arg1 and similar identifiers before? If not, I'd think it'll be fine here too.

Yes, any of those identifiers are sloppy and poor tooling I think. I don't want to repeat that

**Comment by Happypig375** [comment]
If we leave the option for future F# support in expression evaluators, I'd say
```
pipe1@input 
pipe1@stage1
pipe2@stage2
```
Because `@` is already reserved for identifier names.

**Comment by dsyme** [comment]
> Because @ is already reserved for identifier names.

This symbol isn't reserved for identifier names

**Comment by smoothdeveloper** [comment]
@dsyme

> Back pipes will not be supported for this. They should be used very rarely (see my talk "F# Code I love"), and are therefore not a priority for tooling like this.

Sounds a bit like "security through obscurity" to me, and if the engineering efforts are similar to other operators, I'd find it preferable to enable similar debugging experience, even if the operator is not favored.

Improving debugging experience, especially in messy codebases is a different concern from "code should be nice at all times".

@baronfel  / @dsyme

>> is it possible to generate unique names (or close-to-unique?) for the intermediate values to sidestep this?

> We could in theory generate unique names (pipe1-input, pipe1-stage1 etc?), though I do wonder if it would be more confusing.

Despite the shadowing is indeed confusing in that context, I think the generated names won't help, and it will reduce the incentive for debugger infrastructure to ever improve to support shadowed names.

IIRC, in the "locals" tab, it is possible to see all the different symbols, despite they share the same identifier, so I'd be in favour of keeping status quo and tracking the issue upstream on the debugger side, even if things stay as they are.

**Comment by dsyme** [comment]
> I'd find it preferable to enable similar debugging experience, even if the operator is not favored.

Part of the problem is that in 
```fsharp
expr1 <|  expr2
```
The evaluation order is 
1. expr1
2. expr2
3. apply result-of-expr1 to result-of-expr2

This doesn't lead to a great natural stepping/breakpoint sequence. Presumably the steps would cover the following ranges:
1. expr1  (covers expr1)
2. expr2 (covers expr2)
3. apply result-of-expr1 to result-of-expr2  (covers whole expression)

This will result in a lot of leaping back and forth, and when placing a breakpoint it will be somewhat random what you get

In contrast, for
```fsharp
expr1 |>  expr2
```
The evaluation order is 
1. expr1
2. expr2
3. apply result-of-expr2 to result-of-expr1

This gives a simplfied two-step natural stepping/breakpoint sequence down the pipeline:
1. stop before evaluating expr1  (covers expr1)
2. stop before evaluating expr2 and applying result-of-expr2 to result-of-expr1  (covers expr2)

This is because `x |> f1 |> f2` is sort of more natural in flow

As an aside, in debug code the separation of stages down the pipeline may potentially create more closures, I'm not sure, I will check.  It doesn't necessarily have to but it is important to check.

**Comment by auduchinok** [comment]
> Yes, any of those identifiers are sloppy and poor tooling I think. I don't want to repeat that

These identifiers are already seen in lots of places in functions/patterns when debugging F# code, and I assumed people learn what they are or simply ignore them. Having a similar naming would at least be consistent and, arguably, won't make things worse than they already are.

**Comment by dsyme** [comment]
> Won't make things worse than they already are.

I want to go over the whole lot and make them better :)  That is, I'm very open to discussing what "better" is! :)

**Comment by auduchinok** [comment]
No entirely serious: we could wrap these values into structs that would have a good debugger presentation and a name that would be accessible in evaluations. 🙂

**Comment by dsyme** [comment]
> Not that serious: in debug mode, we could wrap these values into structs that would have a good debugger presentation (explaining what they are) and a name that would be accessible in evaluations. 🙂

Yes in general we should be doing this sort of thing a lot more.

**Comment by Happypig375** [comment]
@dsyme This `warning FS1104: Identifiers containing '@' are reserved for use in F# code generation` implies that using `@` will lead to less collisions with already in-use identifiers?

**Comment by dsyme** [comment]
Here is screen grab of support for `||>` and `|||>` (yes I know there's a missing space in one of the names)


https://user-images.githubusercontent.com/7204669/129426813-79343c7b-dbb8-4e34-b3fd-bb5ddafe04fe.mp4

**Comment by baronfel** [comment]
If you toss a standalone dll up I might be able to debug it in vscode and see what the outputs look like there

**Comment by dsyme** [comment]
@baronfel You mean this?

[a.zip](https://github.com/dotnet/fsharp/files/6985575/a.zip)

**Comment by dsyme** [comment]
And with breakpoint validation


https://user-images.githubusercontent.com/7204669/129427481-bd894e78-0985-41da-8076-26fc1925a39e.mp4

**Comment by dsyme** [comment]
I've checked the increase in IL size and IL local count in the codegen samples we have in EmittedIL tests, and the increases are small - there are no new closures created (the generated code is actually nice and efficient) and there are a few more IL instructions and a few more locals, but nothing overly dramatic - suitable for better debug code.

This feature is actually very close to what we need.  Breakpoint validation needs some automated unit tests, though as you can see from the video it's actually working pretty well (if it didn't I couldn't lay down the breakpoints at all)

**Comment by baronfel** [comment]
Initial look in Ionide is looking pretty good! I can't go to the source directly because of an error in the debugger (which wouldn't be a problem if I were using code I had built, etc, etc), but in the stack frames we can easily see the nice new locals:

image

Here's an example of what I was talking about, how the .Net Core Debugger in VSCode seems to auto-supply member return values in the locals view. This was taken I think in the range expression in the second pipeline `[0..2]`.
image

Midway through the second pipeline in the sample:
image

**Comment by zanaptak** [comment]
I wonder if the pipe symbol itself could work for generated locals, `|> input:` or `Into |>` or something.

File: src/fsharp/Optimizer.fs [hunk 2]
```diff
@@ -3195,6 +3240,104 @@ and OptimizeApplication cenv env (f0, f0ty, tyargs, args, m) =
                    MightMakeCriticalTailcall = mayBeCriticalTailcall
                    Info=ValueOfExpr newExpr }
     
+/// Extract a sequence of pipe-right operations (note the pipe-right operator is left-associative
+/// so we start with the full thing and descend down taking apps off the end first)
+/// The pipeline begins with a |>, ||> or ||>
```
Comment (Happypig375) [inline src/fsharp/Optimizer.fs:3245]:
```suggestion
/// The pipeline begins with a |>, ||> or |||>
```

File: src/fsharp/SyntaxTreeOps.fs (hunk 1 above)
Comment (forki) [reply src/fsharp/SyntaxTreeOps.fs]: Instead of debugging support we could create a refactoring that rewrites backpipes to parens

**Comment by auduchinok** [comment]
> Option 3 - uppercase start, spaces with text
>
> ```
> Pipe #1 input
> Pipe #1 stage #1
> Pipe #1 stage #2
> ```

@dsyme What do you think about having the pipe start line number instead of or in addition to the pipe number? The line where a pipe starts might be a bit more helpful than the number of the pipe.

**Comment by Happypig375** [comment]
And to be consistent with stack traces for lambdas as well as for future usage from F# support in the debugger evaluator, we can use `@` in the shown name to indicate the line number.

**Comment by dsyme** [comment]
> @dsyme What do you think about having the pipe start line number instead of or in addition to the pipe number? The line where a pipe starts might be a bit more helpful than the number of the pipe.

This seems reasonable

> And to be consistent with stack traces for lambdas as well as for future usage from F# support in the debugger evaluator, we can use @ in the shown name to indicate the line number.

Perhaps, it seems ok, though we could consider changing that to something clearer, or use C#-like naming for closure classes

**Comment by baronfel** [comment]
Here's a short gif video of vscode debugging stepping through this feature at the current commit: 

debugsteps

The major takeaway is that things mostly Just Work, with some oddities around how putting breakpoints in the lambdas work. You can see the breakpoints at the start of the line bind to the first expression in the lambda instead of the fancy new points, but that is likelyt o be out of our control in Ionide.  Luckily when you step out the lambda you get the good pipeline locals again.

**Comment by dsyme** [comment]
> Here's a short gif video of vscode debugging stepping through this feature at the current commit:
>
> ...but that is likelyt o be out of our control in Ionide...

Is Ionide using ValidateBreakpointLocation for the language service? If not:

- ValidateBreakpointLocation is used to lay down breakpoints (takes a postion to a range) *before* debugging is launched
- The actual debugger PDB information is used *during* debugging

VBPL should be exposed as part of the language server protocol and AKAIK should be a standard sort of thing for languages to implement in VS Code.

**Comment by baronfel** [comment]
There's a [separate specification for debugging in VSCode](https://microsoft.github.io/debug-adapter-protocol/specification) and it does have a request for [setting a breakpoint](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_SetBreakpoints), but in Ionide we are completely reliant on the executable provided by microsoft that implements this protocol. This is [shipped and activated](https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/coreclr-debug/activate.ts#L132-L185) in Omnisharp and is closed source, so I personally have zero visibility into its capabilities or extensibility.

**Comment by dsyme** [comment]
I've added systematic tests for ValidateBreakpointLocation and this is ready.

**Comment by dsyme** [comment]
> There's a separate specification for debugging in VSCode and it does have a request for setting a breakpoint, but in Ionide we are completely reliant on the executable provided by microsoft that implements this protocol. This is shipped and activated in Omnisharp and is closed source, so I personally have zero visibility into its capabilities or extensibility.

Yes, ultimately we will need our own F# implementation of this - otherwise the debug experience (before launching debug) will be messed up.  It won't be a complex thing - presumably we can look at the TypeScript or other implementations.

File: src/fsharp/Optimizer.fs [hunk 3]
```diff
@@ -295,16 +295,21 @@ let [<Literal>] jitOptDefault = true
...
-    { abstractBigTargets : bool
+    { 
+      abstractBigTargets : bool
       
       jitOptUser : bool option
       
       localOptUser : bool option
       
-      crossModuleOptUser : bool option 
+      debugPointsForPipeRight: bool option
```
Comment (TIHan) [inline src/fsharp/Optimizer.fs:310]: We are handling debug points in a particular way for `PipeRight` based on this recd field. I guess it's far out of scope to generalize anything and not worth that amount of effort - so it makes sense.

File: src/fsharp/Optimizer.fs [hunk 4]
```diff
@@ -3195,6 +3240,104 @@ and OptimizeApplication cenv env (f0, f0ty, tyargs, args, m) =
...
+    // turn into let-binding on an intermediate pipe stage
+    //    let pipe-stage-n = fMid <prev-pipe-input> 
+    //    rest <pipe-stage-n>
+    // with a breakpoint on the binding
+    //
+    let pipesBinder =
+        List.foldBack 
+            (fun (i, (xsRange, resType, fExpr: Expr, _)) binder ->
+                let name = $"Pipe #%d{env.methEnv.pipelineCount} stage #%d{i+1}"
+                let stageVal, stageValExpr = mkLocal (List.reduce unionRanges xsRange) name resType
```
Comment (TIHan) [inline src/fsharp/Optimizer.fs:3300]: Makes sense - you have to store the result of the value as a local if we are even to see the result when stepping. Also can become beneficial if you want to see the chain of how the value transformed over the course of the pipes.

File: src/fsharp/Optimizer.fs [hunk 5]
```diff
@@ -397,10 +422,17 @@ type cenv =
       
       /// cache methods with SecurityAttribute applied to them, to prevent unnecessary calls to ExistsInEntireHierarchyOfType
       casApplied: Dictionary<Stamp, bool>
+
     }
 
     override x.ToString() = "<cenv>"
 
+// environment for a method
+type menv =
```
Comment (TIHan) [inline src/fsharp/Optimizer.fs:431]: nit: I think we should rename this to `methEnv` or `MethodEnv` considering there are a lot of `(insert letter)env` types; would make it less ambiguous

**Review by TIHan — APPROVED** [review]
Looks like a big change, but it's relatively small with a large amount of impact. :) - Most of the changes in the PR are updating the baseline tests.

I only had a minor nit on the name `menv`.

**Pull request merged**